### PR TITLE
Fix `dbg_xlate_linear2phy` for NPT

### PR DESCRIPTION
**Pull request by tandasat**
`dbg_xlate_linear2phy()` does not translate the guest physical address for a AMD processor. This prevents the Bochs debugger from displaying memory contents and disassembled code when the processor is running in the guest mode, making it quite useless.

This patch fixes this by enabling translation by calling `dbg_translate_guest_physical` . `dbg_translate_guest_physical` was updated to get nCr3 and check PTEs with a different bitmask when the translation is being done for AMD. 

Focus on the 2nd commit for review.

### CODE REVIEW

File: bochs/cpu/paging.cc [hunk 1]
```diff
@@ -2320,6 +2350,14 @@ bool BX_CPU_C::dbg_xlate_linear2phy(bx_address laddr, bx_phy_address *phy, bx_ad
   if (! long_mode()) laddr &= 0xffffffff;
 #endif
 
+  bool translate_guest_physical = false;
+#if BX_SUPPORT_VMX >= 2
+  translate_guest_physical |= (BX_CPU_THIS_PTR in_vmx_guest && SECONDARY_VMEXEC_CONTROL(VMX_VM_EXEC_CTRL3_EPT_ENABLE));
+#endif
+#if BX_SUPPORT_SVM
+  translate_guest_physical |= (BX_CPU_THIS_PTR in_svm_guest && SVM_NESTED_PAGING_ENABLED);
+#endif
```
Comment (tandasat) [inline bochs/cpu/paging.cc:2359]: This refactoring was done to avoid duplicating those condition checks 3 times. 
```c
#if BX_SUPPORT_VMX >= 2
   // (Condition check for VMX) {
     if (! dbg_translate_guest_physical(pt_address, &pt_address, verbose))
       goto page_fault;
  }
#endif
#if BX_SUPPORT_SVM
   // (Condition check for SVM) {
     if (! dbg_translate_guest_physical(pt_address, &pt_address, verbose))
       goto page_fault;
  }
#endif
```

File: bochs/cpu/cpu.h [hunk 2]
```diff
@@ -4292,9 +4292,7 @@ class BOCHSAPI BX_CPU_C : public logfunctions {
   BX_SMF bool  dbg_instruction_epilog(void);
 #endif
   BX_SMF bool  dbg_xlate_linear2phy(bx_address linear, bx_phy_address *phy, bx_address *lpf_mask = 0, bool verbose = 0);
-#if BX_SUPPORT_VMX >= 2
   BX_SMF bool dbg_translate_guest_physical(bx_phy_address guest_paddr, bx_phy_address *phy, bool verbose = 0);
-#endif
```
Comment (tandasat) [inline bochs/cpu/cpu.h:4297]: Now this function is compiled unconditionally. It is never called if neither SVM or VMX>2 enabled.

File: bochs/cpu/paging.cc [hunk 3]
```diff
@@ -2267,28 +2267,56 @@ void dbg_print_ept_paging_pte(int level, Bit64u entry, bool mbe)
...
+      case BX_EPT_ENTRY_WRITE_ONLY:
+      case BX_EPT_ENTRY_WRITE_EXECUTE:
+        return 0;
+      }
+    }
+#endif
+#if BX_SUPPORT_SVM
+    if (! vmx) {
+      if (!(pte & 1))
+        return 0;
```
Comment (tandasat) [inline bochs/cpu/paging.cc:2317]: The layout of the NTP entry is identical to the standard PTE. Meaning that only when missing the bit 0, it is unreadable.

**Review by stlintel — COMMENTED** [review]
Please split non-functional changes (like whitespace cleanup) from functional PR

File: bochs/cpu/cpu.h (hunk 2 above)
Comment (stlintel) [reply bochs/cpu/cpu.h:4297]: May be better to compile it if VMX >= 2 || SVM

File: bochs/cpu/paging.cc [hunk 4]
```diff
@@ -2308,8 +2336,10 @@ bool BX_CPU_C::dbg_translate_guest_physical(bx_phy_address guest_paddr, bx_phy_a
 
   *phy = pt_address + (bx_phy_address)(guest_paddr & offset_mask);
   return 1;
+#else
```
Comment (stlintel) [inline bochs/cpu/paging.cc:2339]: if compiled for VMX || SVM there will be no else

**Review by stlintel — CHANGES_REQUESTED** [review]

File: bochs/cpu/paging.cc (hunk 4 above)
Comment (tandasat) [reply bochs/cpu/paging.cc]: Not sure if I followed but the code around it has changed.

**Comment by tandasat** [comment]
> Please split non-functional changes (like whitespace cleanup) from functional PR

I dropped the commit.

Can you explain what is the reason for this rule? IMO, this should be allowed, as long as such changes are separated from functional change commits..

To my knowledge, not leaving trailing spaces is a common practice, and so, many of developers configure their editors to auto-delete trailing spaces. This rule (not mixing functional & non-functional changes) will be annoyance to contribute to this projects for many, and passively encourage contributors leave trailing spaces (for example, I am not going to create a PR for only removing trailing spaces).
  
GitHub allows reviewing select commits, and we can revert functional-commits and leave trailing space removal commits if needed. I do not see much benefit of the rule.

File: bochs/cpu/cpu.h (hunk 2 above)
Comment (tandasat) [reply bochs/cpu/cpu.h:4297]: Resolved in [0bcfcb0](https://github.com/stlintel/Bochs/pull/30/commits/0bcfcb062c22de6984bfa07ae906390beccb8cb2)

**Comment by stlintel** [comment]
It is very hard to follow what are the functional changes and what are just whitespace.
If there are small amount of whitespace changes it is easy to follow but with 9/10 of the PR - it is very inconvenient.
You could just submit single PR with whitespace cleanups in entire CPU code if you want, it will be accepted.

**Comment by stlintel** [comment]
Actually your entire patch is under assumption that SVM is using long mode PAE mode for nested walk
Which is incorrect assumption.
Nested walk could be long mode, PAE or even legacy walk.
At the end I suggest to drop the PR and rewrite it properly to support all SVM vases or at least print nicely when dbg walk is not supported / not implemented yet

**Comment by stlintel** [comment]
BTW - you may notice that nested walk is just the same as regular walk in x86 just starting from another base
So might be the best way is to reuse existing dbg_xlate_linear2phy may be adding a parameter to it saying it is nested walk
The parameter will prevent even deeper nested walk and do smarter verbose prints.

**Comment by tandasat** [comment]
Good catch on the paging mode, and thanks for the idea of reusing dbg_xlate_linear2phy. I will look into it and/or showing warning. I may scrap this PR and recreate a new one, or may just remove previous commits.

I might create the mass trailing space removal PR for `cpu`, or even better, all source files if that is acceptable.

**Comment by therealdreg** [comment]
> I might create the mass trailing space removal PR for `cpu`, or even better, all source files if that is acceptable.

If we will change all the source code for cosmetic reasons, (btw I'm not sure if this is a good idea). 

We could take this opportunity to force and fix the coding style of the whole project with an automatic tool like clang-tidy. 

clang-tidy does a deeper inspection and can be used to flag things that are "bad practice", not just "bad formatting". This would help to have more consistency.

But as I said, I'm not sure if it is a good idea modify all the source code :\

**Comment by stlintel** [comment]
I thought about it but I am not a master to configure the tool anyway.
Also forcing people to specific very strict style is not a way to go in my opinion.
But obvious bad style should be fixed.

File: bochs/cpu/paging.cc [hunk 5]
```diff
@@ -2417,11 +2442,17 @@ bool BX_CPU_C::dbg_xlate_linear2phy(bx_address laddr, bx_phy_address *phy, bx_ad
 #if BX_SUPPORT_VMX >= 2
   if (BX_CPU_THIS_PTR in_vmx_guest) {
     if (SECONDARY_VMEXEC_CONTROL(VMX_VM_EXEC_CTRL3_EPT_ENABLE)) {
-      if (! dbg_translate_guest_physical(paddress, &paddress, verbose))
+      if (! dbg_translate_guest_physical_with_ept(paddress, &paddress, verbose))
```
Comment (stlintel) [inline bochs/cpu/paging.cc:2445]: Looks nice, just some naming ...

Could you rename 
dbg_translate_guest_physical_with_ept -> dbg_translate_guest_physical_ept
dbg_translate_guest_physical_with_npt -> dbg_translate_guest_physical_npt
second_level_translation => nested_walk

**Review by stlintel — CHANGES_REQUESTED** [review]

**Comment by tandasat** [comment]
Update the PR to reuse `dbg_xlate_linear2phy`. Dropped previous commits.

Mass deletion of trailing spaces should not be really controversial. Setting up something to prevent addition of trailing spaces (eg, `.editorconfig`) and to enforce other coding styles (eg, `.clang-format`) are. Let us create their own issues and discuss there.

File: bochs/cpu/paging.cc (hunk 5 above)
Comment (tandasat) [reply bochs/cpu/paging.cc:2445]: will do.

Comment (tandasat) [reply bochs/cpu/paging.cc]: Fixed with 179c95d

**Review by stlintel — APPROVED** [review]
